File: CSharp/One/Transforms/ResolveImports.cs
using One.Ast;
using One;

namespace One.Transforms
{
    public class ResolveImports : AstTransformer {
        public Workspace workspace;
        
        public ResolveImports(Workspace workspace): base("ResolveImports")
        {
            this.workspace = workspace;
        }
        
        public override void visitFile(SourceFile sourceFile)
        {
            ResolveImports.processFile(this.workspace, sourceFile);
        }
        
        public static void processFile(Workspace ws, SourceFile file)
        {
            foreach (var imp in file.imports) {
                var impPkg = ws.getPackage(imp.exportScope.packageName);
                var scope = impPkg.getExportedScope(imp.exportScope.scopeName);
                imp.imports = imp.importAll ? scope.getAllExports() : imp.imports.map(x => x is UnresolvedImport unrImp ? scope.getExport(unrImp.name) : x);
                file.addAvailableSymbols(imp.imports);
            }
        }
        
        public static void processWorkspace(Workspace ws)
        {
            foreach (var pkg in Object.values(ws.packages))
                foreach (var file in Object.values(pkg.files))
                    ResolveImports.processFile(ws, file);
        }
    }
}

File: PHP/One/Transforms/ResolveImports.php
<?php

namespace One\Transforms\ResolveImports;

use One\Ast\Types\Workspace;
use One\Ast\Types\UnresolvedImport;
use One\Ast\Types\SourceFile;
use One\Ast\Types\Package;
use One\AstTransformer\AstTransformer;

class ResolveImports extends AstTransformer {
    public $workspace;
    
    function __construct($workspace) {
        parent::__construct("ResolveImports");
        $this->workspace = $workspace;
    }
    
    function visitFile($sourceFile) {
        ResolveImports::processFile($this->workspace, $sourceFile);
    }
    
    static function processFile($ws, $file) {
        foreach ($file->imports as $imp) {
            $impPkg = $ws->getPackage($imp->exportScope->packageName);
            $scope = $impPkg->getExportedScope($imp->exportScope->scopeName);
            $imp->imports = $imp->importAll ? $scope->getAllExports() : array_map(function ($x) use ($scope) { return $x instanceof UnresolvedImport ? $scope->getExport($x->name) : $x; }, $imp->imports);
            $file->addAvailableSymbols($imp->imports);
        }
    }
    
    static function processWorkspace($ws) {
        foreach (array_values($ws->packages) as $pkg)
            foreach (array_values($pkg->files) as $file)
                ResolveImports::processFile($ws, $file);
    }
}
